proxyclient/sprr_test_permissions.py

```diff
@@ -0,0 +1,386 @@
+
+class ARMPageTable:
+	PAGESIZE = 0x4000
+
+	def __init__(self, memalign, free):
+		self.memalign = memalign
+		self.free = free
+
+		self.l0 = self.memalign(self.PAGESIZE, self.PAGESIZE)
+		self.l1 = [self.memalign(self.PAGESIZE, self.PAGESIZE), self.memalign(self.PAGESIZE, self.PAGESIZE)]
+		self.l2 = {}
+
+		p.write64(self.l0, self.make_table_pte(self.l1[0]))
+		p.write64(self.l0+8, self.make_table_pte(self.l1[1]))
+
+
+	def make_table_pte(self, addr):
+		# table mapping, access bit set
+		return addr | 0b11 | (1<<10)
+
+	def map_page(self, vaddr, paddr, access_bits):
+		ap  = (access_bits & 0b1100) >> 2
+		pxn = (access_bits & 0b0010) >> 1
+		uxn = (access_bits & 0b0001)
+
+		# block mapping, access bit set
+		pte = paddr | 0b01 | (1<<10)
+
+		# move access bits in place
+		pte |= ap << 6
+		pte |= pxn<<54
+		pte |= uxn<<53
+
+		l0_idx = (vaddr>>(25+11+11)) & 1
+		l1_idx = (vaddr>>(25+11)) & 0x7ff
+		l2_idx = (vaddr>>25) & 0x7ff
+
+		tbl = self.l2.get((l0_idx, l1_idx), None)
+		if not tbl:
+			tbl = self.memalign(self.PAGESIZE, self.PAGESIZE)
+			self.l2[(l0_idx, l1_idx)] = tbl 
+			p.write64(self.l1[l0_idx] + 8*l1_idx, self.make_table_pte(tbl))
+	
+		p.write64(tbl + 8*l2_idx, pte)
+	
+	def map(self, vaddr, paddr, sz, access_bits):
+		assert sz % 0x2000000 == 0
+		assert vaddr % 0x2000000 == 0
+		assert paddr % 0x2000000 == 0
+		assert access_bits <= 0b1111
+
+		while sz > 0:
+			self.map_page(vaddr, paddr, access_bits)
+			sz -= 0x2000000
+			vaddr += 0x2000000
+			paddr += 0x2000000
+
+
+def build_and_write_code(heap, code):
+        page = heap.memalign(0x4000, 0x4000)
+        compiled = asm.ARMAsm(code, page).data
+        iface.writemem(page, compiled)
+        p.dc_cvau(page, len(compiled))
+        p.ic_ivau(page, len(compiled))
+        return page
+
+def setup_exception_vectors(heap, gxf=False):
+        if gxf:
+                elr = "S3_6_C15_C10_6"
+                eret = ".long 0x201400"
+                indicator = 0xf2
+        else:
+                elr = "ELR_EL1"
+                eret = "eret"
+                indicator = 0xf0
+
+        return build_and_write_code(heap, """
+                .rept   16
+                b 1f
+                .align 7
+                .endr
+
+                1:
+                // store that we failed
+                mov x10, 0x{indicator:x}
+
+                // move PC two instruction further and clear 0xf0 0000 0000 to
+                // make sure we end up in the r-x mapping either way and don't
+                // repeat the instruction that just faulted
+                // we skip the second instruction since that one is used to
+                // indicate success
+                ldr x11, =0x0fffffffff
+                mrs x12, {elr}
+                add x12, x12, #8
+                and x12, x12, x11
+                msr {elr}, x12
+
+                isb
+                {eret}
+        """.format(eret=eret, elr=elr, indicator=indicator))
+
+with guarded_heap() as heap:
```

Comment:
... why are you using it here? The utils heap is managed Python-side, so everything vanishes after the script exits, which is why currently almost nothing here cares about cleaning up allocations.

This context manager will probably be useful in the future, once we start e.g. using the hypervisor which will be a long-running script and may need to manage more complex allocation situations.

Reply:
Because this code originally looked quite different and the guarded heap was a leftover I forgot to remove :)